src/sourcekit-lsp/getReferenceDocument.ts

```diff
@@ -0,0 +1,45 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the VS Code Swift open source project
+//
+// Copyright (c) 2024 the VS Code Swift project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of VS Code Swift project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import * as vscode from "vscode";
+import * as langclient from "vscode-languageclient/node";
+import { GetReferenceDocumentParams, GetReferenceDocumentRequest } from "./lspExtensions";
+
+export function activateGetReferenceDocument(client: langclient.LanguageClient): vscode.Disposable {
+    const getReferenceDocument = vscode.workspace.registerTextDocumentContentProvider(
+        "swift-macro-expansion",
+        {
+            provideTextDocumentContent: async (uri, token) => {
+                const params: GetReferenceDocumentParams = {
+                    uri: uri.toString(true),
+                };
+
+                const expansion = await client.sendRequest(
+                    GetReferenceDocumentRequest,
+                    params,
+                    token
+                );
+
+                if (expansion) {
+                    return expansion.content;
+                } else {
+                    vscode.window.showWarningMessage("Unable to retrieve Macro Expansion");
```

Comment:
Instead of showing a warning message, perhaps just return the contents of the document as `Unable to retrieve reference document ...`